### PR TITLE
Separate DNS certificate from other certificates

### PR DESCRIPTION
**Pull request by lei-tang**
- Separate DNS certificate from other certificates (e.g., SPIFFE, SDS, etc) so each certificate can be used for a specific purpose.
- Certificate controller is created before MCP controller in case MCP server pod waits to mount a certificate to be provisioned by the certificate controller.

Please provide a description for what this PR is for: https://github.com/istio/istio/issues/17061

Design documents:
[1]https://docs.google.com/document/d/1VHyBre8943USOx_YEVtA18KfEoGTmLT1iBHESO5bzcA/edit#heading=h.qex63c29z2to
[2] https://docs.google.com/document/d/1v8BxI07u-mby5f5rCruwF7odSXgb9G8-C9W5hQtSIAg/edit#heading=h.xw1gqgyqs5b

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/bootstrap/server.go [hunk 1]
```diff
@@ -276,6 +276,13 @@ func NewServer(args PilotArgs) (*Server, error) {
 	if err := s.initMeshNetworks(&args); err != nil {
 		return nil, fmt.Errorf("mesh networks: %v", err)
 	}
+	// Certificate controller must be created before MCP
+	// controller to avoid deadlocks caused by the circular dependency, i.e.,
+	// MCP server pod waits to mount a certificate while
+	// Pilot waits MCP server.
```
Comment (rlenglet) [inline pilot/pkg/bootstrap/server.go]:
```suggestion
	// Pilot waits for MCP server.
```

**Review by rlenglet — APPROVED** [review]

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/bootstrap/server.go (hunk 1 above)
Comment (lei-tang) [reply pilot/pkg/bootstrap/server.go]: Done.

File: pilot/pkg/bootstrap/server.go [hunk 2]
```diff
@@ -276,6 +276,13 @@ func NewServer(args PilotArgs) (*Server, error) {
 	if err := s.initMeshNetworks(&args); err != nil {
 		return nil, fmt.Errorf("mesh networks: %v", err)
 	}
+	// Certificate controller must be created before MCP
+	// controller to avoid deadlocks caused by the circular dependency, i.e.,
+	// MCP server pod waits to mount a certificate while
+	// Pilot waits for MCP server.
+	if err := s.initCertController(&args); err != nil {
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/server.go]: I understand the init here just create a controller, what really matters is the start func `s.certController.Run(stop)`.

Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/server.go]: I mean in order to speed up istio bootstrap, could we create the necessary certs synchronously, and leave the reconcile in the `certController.Run` which is asynchronously started.

Comment (lei-tang) [reply pilot/pkg/bootstrap/server.go]: The subject and the implementation of this PR has been revised: Separate DNS certificate from other certificates (e.g., SPIFFE, SDS, etc) so each certificate can be used for a specific purpose.

Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/server.go]: Ah, noticed

File: pilot/pkg/bootstrap/server.go [hunk 3]
```diff
@@ -276,6 +276,13 @@ func NewServer(args PilotArgs) (*Server, error) {
 	if err := s.initMeshNetworks(&args); err != nil {
 		return nil, fmt.Errorf("mesh networks: %v", err)
 	}
+	// Certificate controller must be created before MCP
+	// controller to avoid deadlocks caused by the circular dependency, i.e.,
+	// MCP server pod waits to mount a certificate while
+	// Pilot waits for MCP server.
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/server.go]: Should also update the comment

Comment (lei-tang) [reply pilot/pkg/bootstrap/server.go]: Done: comment has been updated.

File: install/kubernetes/helm/istio/charts/galley/templates/deployment.yaml [hunk 4]
```diff
@@ -84,11 +84,21 @@ spec:
 {{- end}}
 {{- if .Values.enableAnalysis }}
           - --enableAnalysis=true
+{{- end }}
+{{- if .Values.global.certificates }}
+          - --validation.tls.clientCertificate=/etc/dnscerts/cert-chain.pem
+          - --validation.tls.privateKey=/etc/dnscerts/key.pem
+          - --validation.tls.caCertificates=/etc/dnscerts/root-cert.pem
```
Comment (ayj) [inline install/kubernetes/helm/istio/charts/galley/templates/deployment.yaml]: This presumes the list always includes the galley DNS cert. Maybe check if `dns.istio-galley-service-account` is in the list?

Comment (lei-tang) [reply install/kubernetes/helm/istio/charts/galley/templates/deployment.yaml]: It is not straightforward for helm to parse .Values.global.certificates, understanding its proto structure and doing the check. Thus, in current implementation, similar to the approach for configuring other helm values, the user configuring .Values.global.certificates needs to ensure the value is valid.

**Review by ayj — APPROVED** [review]

**Pull request merged**